src/Native/System.Private.CoreLib.Native/pal_threading.cpp

```diff
@@ -0,0 +1,134 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "pal_common.h"
+#include "pal_threading.h"
+#include "pal_time.h"
+
+#include <sched.h>
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// LowLevelMutex
+
+extern "C" void CoreLibNative_LowLevelMutex_Acquire(LowLevelMutex *mutex)
+{
+    assert(mutex != nullptr);
+    mutex->Acquire();
+}
+
+extern "C" void CoreLibNative_LowLevelMutex_Release(LowLevelMutex *mutex)
+{
+    assert(mutex != nullptr);
+    mutex->Release();
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// LowLevelMonitor
+
+#if !(HAVE_MACH_ABSOLUTE_TIME || HAVE_PTHREAD_CONDATTR_SETCLOCK && HAVE_CLOCK_MONOTONIC)
+#error Don't know how to perfom timed wait on this platform
+#endif
+
+LowLevelMonitor::LowLevelMonitor()
+{
+#if HAVE_MACH_ABSOLUTE_TIME
+    // Older versions of OSX don't support CLOCK_MONOTONIC, so we don't use pthread_condattr_setclock. See
+    // Wait(int32_t timeoutMilliseconds).
+    int initError = pthread_cond_init(&m_condition, nullptr);
+#else
+    int error;
+
+    pthread_condattr_t conditionAttributes;
+    error = pthread_condattr_init(&conditionAttributes);
+    if (error != 0)
+    {
+        throw OutOfMemoryException();
```

Comment:
We should not be using C++ exceptions in the CoreLib PAL. It creates dependency on C++ runtime that is a pain. None of the other System.*.Native or CoreRT runtime are using C++ exceptions.